rook/main.py

```diff
@@ -0,0 +1,278 @@
+  for filename in os.listdir(directory):
+    if filename.endswith(".py") and not filename.startswith("__"):
+      module = __import__(filename[:-3]) #[:-3] to remove .py
+      for name, value in module.__dict__.items():
+        #print("Unknown", name, value)
+        if inspect.isclass(value) and value is not Tester\
+           and issubclass(value, Tester):
+          tester_dict[name] = value
+        if inspect.isclass(value) and value is not Differ\
+           and issubclass(value, Differ):
+          differ_dict[name] = value
+
+  return tester_dict, differ_dict
+
+def sec_format(runtime):
+  """
+  Formats the runtime into a string of the number seconds.
+  If runtime is none, format as None!
+  runtime: float or None, runtime to be formated.
+  return str of runtime.
+  """
+  if isinstance(runtime, float):
+    return "{:6.2f}sec".format(runtime)
+  return "  None!  "
+
+def process_result(index, _input_data, output_data):
+  """
+  This is a callback function that Processes the result of a test.
+  index: int, Index into functions list.
+  _input_data: the input data passed to the function
+  output_data: the output data passed to the function
+  """
+  bucket = output_data.bucket
+  process_test_name = test_name_list[index]
+  if bucket == Tester.bucket_success:
+    results["pass"] += 1
+    for postreq in function_postreq.get(process_test_name, []):
+      if postreq in name_to_id:
+        job_id = name_to_id[postreq]
+        print("Enabling", postreq, job_id)
+        run_pool.enable_job(job_id)
+  elif bucket == Tester.bucket_skip:
+    results["skipped"] += 1
+    print(output_data.message)
+  else:
+    results["fail"] += 1
+    failed_list.append(process_test_name)
+    print(output_data.output)
+    print(output_data.message)
+  number_done = sum(results.values())
+  print("({}/{}) {:7s} ({}) {}".format(number_done, len(function_list),
+                                       Tester.get_bucket_name(bucket),
+                                       sec_format(output_data.runtime),
+                                       process_test_name))
+if __name__ == "__main__":
+
+  test_re = re.compile(args.test_re_raw)
+
+  #XXX fixme to find a better way to the tests directory
+
+  this_dir = os.path.abspath(os.path.dirname(__file__))
+  up_one_dir = os.path.dirname(this_dir)
+  base_test_dir = os.path.join(up_one_dir, "tests")
+
+  print(this_dir, base_test_dir)
+
+
+  test_list = get_test_lists(base_test_dir)
+
+  base_testers, base_differs = get_testers_and_differs(this_dir)
+  testers, differs = get_testers_and_differs(os.path.join(up_one_dir, "scripts",
+                                                          "TestHarness", "testers"))
+  testers.update(base_testers)
+  differs.update(base_differs)
+
+  print("Testers:", testers)
+  print("Differs:", differs)
+
+  tester_params = {}
+  for tester in testers:
+    tester_params[tester] = testers[tester].get_valid_params()
+  print("Tester Params:", tester_params)
```

Comment:
Maybe in the future, this could be formatted to print prettier than dicts usually print.

Reply:
I improved the printing (but still could probably be improved more) and changed it so it only prints if ```--list-testers``` is used.